File: java/0108-convert-sorted-array-to-binary-search-tree.java
class Solution {
    public TreeNode sortedArrayToBST(int[] nums) {
        return generateTree(nums, 0, nums.length - 1);
    }

    public TreeNode generateTree(int[] nums, int low, int high) {
        if (low > high) {
            return null;
        }

        int mid = low + ((high - low) / 2);
        TreeNode node = new TreeNode(nums[mid]);

        node.left = generateTree(nums, low, mid - 1);
        node.right = generateTree(nums, mid + 1, high);

        return node;
    }
}

File: kotlin/0108-convert-sorted-array-to-binary-search-tree.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun sortedArrayToBST(nums: IntArray): TreeNode? {
        
        fun createTree(left: Int, right: Int): TreeNode? {

            if(left > right) return null
            else if(left == right) return TreeNode(nums[left])

            val mid = (left + right) / 2
            val node = TreeNode(nums[mid])
            node.left = createTree(left, mid-1)
            node.right = createTree(mid+1, right)

            return node
        }

        return createTree(0, nums.lastIndex)
    }
}
